content/blog/how-tikv-accesses-data.md

```diff
@@ -0,0 +1,263 @@
+
+## TxnKV
+
+TxnKV corresponds to the Percolator mentioned above. It is a higher-level key-value API that provides ACID semantics. TxnKV is used when your application requires distributed transactions and/or MVCC.
+
+
+### Write with TxnKV
+
+{{< figure
+    src="/img/blog/how-tikv-accesses-data/txnkv-write.png"
+    caption="Write with TxnKV"
+    number="" >}}
+
+For TxnKV, the situation is much more complicated. Most of the processes have been explained in the [Percolator](#percolator) section. For TiKV to quickly seek the latest commit, the following steps are performed:
+
+1. Use big-endian to convert the timestamp to a 8-byte value, and then invert the bytes by bit.
+2. Combine the inverted bytes with the original key for storage in RocksDB
+
+This way the latest commits can be stored in the front, and the seek can be directly located. Of course, the corresponding TS is also encoded and processed in the same way.
+
+For example, suppose a key now has two commits—commitTS is 10 and 12, and startTS is 9 and 11, then the order of keys in RocksDB is:
+
+```
+Write CF:
+
+a_12-> 11
+a_10-> 9
+
+Data CF:
+
+a_11-> data_11
+a_9-> data_9
+```
+
+In addition, it should be noted that for small values, TiKV directly stores the value in the Write CF, so that only Read CF is required when reading. For write operations, the process is as follows:
+
+```
+PreWrite:
+
+Lock CF: W a-> Lock + Data
+
+Commit:
+Lock CF: R a-> Lock + 10 + Data
+Lock CF: D a
+
+Write CF: W a_11-> 10 + Data
+```
+
+For TiKV, the Lock CF will be read during the Commit phase to determine transaction conflicts, so we can get data from the Lock CF and write it to the Write CF.
+
+### Read with TxnKV
+
+{{< figure
+    src="/img/blog/how-tikv-accesses-data/txnkv-read.png"
+    caption="Read with TxnKV"
+    number="" >}}
+
+Refer to the [Percolator](#percolator) section for detailed write process with the TxnKV.
+
+### SQL Key Mapping between TiDB and TiKV
+
+TiKV is a distributed KV store. On top of it we built TiDB, a distributed relational database. You may wonder how a relational table is mapped to key-value. Take the following table:
+
+```
+CREATE TABLE t1 {
+	id BIGINT PRIMARY KEY,
+	name VARCHAR (1024),
+	age BIGINT,
+	content BLOB,
+	UNIQUE (name),
+	INDEX (age),
+}
+```
+
+In this example, we create table t1 with 4 fields, with `ID` as the primary key, `name` as the unique index, and `age` as a non-unique index. So how does the data in this table correspond to TiKV?
+
+In TiDB, each table has a unique ID, such as 11 here, and each index also has a unique ID, for example, 12 for the name index and 13 for the age index. We use prefixes `t` and `i` to distinguish between data and index in the table. For table `t1` above, suppose it now has two rows of data, which are (1, “a”, 10, “hello”) and (2, “b”, 12, “world”). In TiKV, each row of data has different a corresponding key-value pair, as shown below:
+
+```
+PK
```

Comment:
What "PK" is may not be immediately clear to beginner readers, suggest expanding it.